### PR TITLE
Supports all static image types retrieved from UIImagePickerController

### PR DESCRIPTION
**Pull request by esreli**
Implements adding photo attachment using [addAttachmentWithUIImagePickerControllerInfoDictionary](https://developers.arcgis.com/ios/latest/api-reference/interface_a_g_s_popup_attachment_manager.html#a23bfef7aaa1e99770ccc6c9566cb4ade) rather than [addAttachmentAsJPGWithImage](https://developers.arcgis.com/ios/latest/api-reference/interface_a_g_s_popup_attachment_manager.html#a9606cea7ec9b1a1d20cd5da1e52e15a4).

See https://github.com/Esri/data-collection-ios/issues/171

### CODE REVIEW

File: data-collection/data-collection/Rich Popup/Attachments/RichPopupAttachmentsManager.swift [hunk 1]
```diff
@@ -139,7 +145,16 @@ class RichPopupAttachmentsManager: AGSLoadableBase {
             }
         }
         
-        discardStagedAttachments()
+        dispatchGroup.notify(queue: OperationQueue.current?.underlyingQueue ?? .main) { [weak self] in
```
[the remark this answers is not in the document]
Comment (esreli) [reply data-collection/data-collection/Rich Popup/Attachments/RichPopupAttachmentsManager.swift]: As far as i'm aware, this wouldn't work. According to the [docs](https://developer.apple.com/documentation/dispatch/dispatchgroup/2016066-notify):

> If the group is empty (no block objects are associated with the dispatch group), the notification block object is submitted immediately.

[the remark this answers is not in the document]
Comment (philium) [reply data-collection/data-collection/Rich Popup/Attachments/RichPopupAttachmentsManager.swift]: Ah! Did not know that. Thanks!

File: data-collection/data-collection/Rich Popup/Attachments/RichPopupStagedPhotoAttachment.swift [hunk 2]
```diff
@@ -25,45 +26,23 @@ class RichPopupStagedPhotoAttachment: RichPopupStagedAttachment {
...
     
-    convenience init?(imagePickerMediaInfo info: [UIImagePickerController.InfoKey : Any]) {
+     init?(imagePickerMediaInfo info: [UIImagePickerController.InfoKey : Any]) {
         
-        guard let image = info[.editedImage] as? UIImage ?? info[.originalImage] as? UIImage else {
-            return nil
-        }
+        self.info = info
+
+        guard (info[.editedImage] as? UIImage ?? info[.originalImage] as? UIImage) != nil else { return nil }
```
[the remark this answers is not in the document]
Comment (esreli) [reply data-collection/data-collection/Rich Popup/Attachments/RichPopupStagedPhotoAttachment.swift]: Tomato tomato ¯\_(ツ)_/¯

File: data-collection/data-collection/Rich Popup/Attachments/RichPopupStagedPhotoAttachment.swift [hunk 3]
```diff
@@ -25,45 +26,23 @@ class RichPopupStagedPhotoAttachment: RichPopupStagedAttachment {
...
         
-        self.init(image: image)
+        super.init(data: Data(), mimeType: "<will-be-inferred>", name: nil)
```
[the remark this answers is not in the document]
Comment (esreli) [reply data-collection/data-collection/Rich Popup/Attachments/RichPopupStagedPhotoAttachment.swift]: That approach solves for `RichPopupStagedPhotoAttachment` but not for `RichPopupStagedAttachment`. I'd like to give our users the ability to easily add other attachment types, if necessary and making those parameters optional reduces the credibility of the class, IMO.

File: data-collection/data-collection/Rich Popup/Attachments/RichPopupStagedPhotoAttachment.swift [hunk 4]
```diff
@@ -13,6 +13,7 @@
 // limitations under the License.
 
 import ArcGIS
+import Photos
```
[the remark this answers is not in the document]
Comment (esreli) [reply data-collection/data-collection/Rich Popup/Attachments/RichPopupStagedPhotoAttachment.swift]: I had considered diving in with `PHAsset`. Forgot to remove, thank you.

File: data-collection/data-collection/Rich Popup/Attachments/RichPopupAttachmentsManager.swift [hunk 5]
```diff
@@ -117,18 +117,24 @@ class RichPopupAttachmentsManager: AGSLoadableBase {
...
     
-    func commitStagedAttachments() {
+    func commitStagedAttachments(_ completion: @escaping () -> Void) {
         
         // Add Staged Attachments
         let newAttachments = self.stagedAttachments
         
+        let dispatchGroup = DispatchGroup()
+        
+        dispatchGroup.enter()
```
[the remark this answers is not in the document]
Comment (esreli) [reply data-collection/data-collection/Rich Popup/Attachments/RichPopupAttachmentsManager.swift]: Perhaps I wanted to account for the possibility of a race condition but upon second look I don't think I have anything to worry about.

File: data-collection/data-collection/Rich Popup/RichPopupManager.swift [hunk 6]
```diff
@@ -133,23 +133,32 @@ class RichPopupManager: AGSPopupManager {
...
-        // Finally, the manager finishes editing it's attributes.
-        super.finishEditing { (error) in
-            
-            if let error = error {
-                relatedRecordsErrors.append(error)
-            }
-
-            if !relatedRecordsErrors.isEmpty {
-                completion(RichPopupManagerError.invalidPopup(relatedRecordsErrors))
+        func finish() {
```
[the remark this answers is not in the document]
Comment (esreli) [reply data-collection/data-collection/Rich Popup/RichPopupManager.swift]: Do you have any suggestions for improvement?

[the remark this answers is not in the document]
Comment (philium) [reply data-collection/data-collection/Rich Popup/RichPopupManager.swift]: One suggestion would be to break `finishEditing(completion:)` into smaller methods as it is currently almost 100 lines. It looks like the contents of that first `if` statement could be extracted into a method that returns an array of errors. Same with that second `if` statement. Maybe the `finish()` function could be turned into an instance method, something like `superDidFinishEditing(accumulatedErrors: [Error])`.

**Comment by esreli** [comment]
One more look @philium ?

**Pull request merged**